elfpy/markets/hyperdrive/hyperdrive_pricing_model.py

```diff
@@ -442,3 +448,282 @@ def calc_tokens_out_given_lp_in(
+        time_stretch: FixedPoint,
+        share_price: FixedPoint,
+        initial_share_price: FixedPoint,
+        minimum_share_reserves: FixedPoint,
+    ) -> FixedPoint:
+        r"""
+        Calculates the maximum amount of shares that can be used to open shorts.
+
+        Parameters
+        ----------
+        share_reserves : FixedPoint
+            The pool's share reserves.
+        bond_reserves : FixedPoint
+            The pool's bonds reserves.
+        longs_outstanding : FixedPoint
+            The amount of longs outstanding.
+        time_stretch : FixedPoint
+            The time stretch parameter.
+        share_price : FixedPoint
+            The share price.
+        initial_share_price : FixedPoint
+            The initial share price.
+
+        Returns
+        -------
+        FixedPoint
+            The maximum amount of shares that can be used to open shorts.
+        """
+        t = ONE_18 - time_stretch
+        price_factor = share_price / initial_share_price
+        k = modified_yield_space_constant(price_factor, initial_share_price, share_reserves, t, bond_reserves)
+        inner_factor = initial_share_price * longs_outstanding / share_price + minimum_share_reserves**t
+        optimal_bond_reserves = (k - price_factor * inner_factor) ** (ONE_18 / t)
+
+        # The optimal bond reserves imply a maximum short of dy = y - y0.
+        return optimal_bond_reserves - bond_reserves
+
+
+class MaxLongResult(NamedTuple):
+    """Result from calculate_max_long."""
+
+    base_amount: FixedPoint
+    bond_amount: FixedPoint
+
+
+# mimic variable from solidity.
+ONE_18 = FixedPoint("1.0")
+
+
+def calculate_max_buy(
+    z: FixedPoint, y: FixedPoint, t: FixedPoint, c: FixedPoint, mu: FixedPoint
+) -> tuple[FixedPoint, FixedPoint]:
+    r"""
+    Calculates the maximum amount of bonds that can be purchased with the specified reserves.
+
+    Parameters
+    ----------
+    z : FixedPoint
+        Amount of share reserves in the pool.
+    y : FixedPoint
+        Amount of bond reserves in the pool.
+    t : FixedPoint
+        Amount of time elapsed since term start.
+    c : FixedPoint
+        Conversion rate between base and shares.
+    mu : FixedPoint
+        Interest normalization factor for shares.
+
+    Returns
+    -------
+    tuple[FixedPoint, FixedPoint]
+        The cost in shares of the maximum bond purchase and the maximum amount of bonds that can be purchased.
+    """
+    # calculate c_div_mu by directly using regular division operator
+    c_div_mu = c / mu
+    k = modified_yield_space_constant(c_div_mu, mu, z, t, y)
+    # calculate optimal_y and optimal_z using regular division and pow operator
+    optimal_y = (k / (c_div_mu + ONE_18)) ** (ONE_18 / t)
+    optimal_z = optimal_y / mu
+
+    # calculate and return the optimal trade sizes by using regular subtraction operator
+    return (optimal_z - z, y - optimal_y)
+
+
+def modified_yield_space_constant(
```

Comment:
We already have a function for this here:
https://github.com/delvtech/elf-simulations/blob/main/elfpy/markets/hyperdrive/yieldspace_pricing_model.py#L701C7-L701C7

They have slightly different parameter arguments. Maybe your intention is to match the solidity arguments? I think this can be done pretty easily by wrapping the existing implementation, or replacing the existing implementation with this one. Either way, it should exist in `yieldspace_pricing_model.py`.